File: tools/ancres-defense.py
#!/usr/bin/env python3
"""Défenses du joueur — ancres des socles, tourelles centrées sur leur pivot.

Depuis l'arbitrage du 05/09, la tourelle d'une défense TOURNE AU RENDU : un
sprite par tourelle au lieu de seize. Deux conséquences, et elles sont
mécaniques, pas esthétiques :

  1. le sprite doit être CARRÉ et centré sur son pivot, sinon la tourelle décrit
     un cercle autour du centre du fichier au lieu de pivoter sur place ;
  2. le côté du carré doit valoir deux fois la distance pivot → pixel le plus
     loin, sinon les canons sortent du cadre à 45° et se font rogner.

Le socle, lui, ne tourne pas : c'est lui qui publie l'ancre où poser la
tourelle, exactement comme une coque de blindé le fait au lot 8. Le détecteur
d'ancre est CELUI DU DÉPÔT, `tools/chassis.py:ancre`, importé tel quel — en
réécrire un second aurait fait deux vérités sur ce qu'est un logement.

    python3 tools/ancres-defense.py
"""
import os
import sys
import json

import numpy as np
from PIL import Image
from scipy import ndimage as nd

RACINE = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, os.path.join(RACINE, 'tools'))
from chassis import ancre                                    # noqa: E402
from joueur_v2 import emprise_par_sprite                     # noqa: E402
from chemins import dossier_sprites                          # noqa: E402

SRC = os.path.join(RACINE, 'art', 'sources')
ANCRES = dossier_sprites('ancres-defense.json')

# ⚠ CHOISI À L'ŒIL, À 40 PX, PAS DÉDUIT. Caler l'embase sur l'OUVERTURE du
# logement est géométriquement juste et visuellement mort : à la taille du jeu,
# le canon disparaît et il ne reste qu'un anneau de couleur autour d'un trou
# noir — on ne distingue plus une Casemate d'un Créneau autrement que par la
# teinte. À 1,6 l'embase déborde sur le rebord, le canon se lit, et les quatre
# coins colorés restent à découvert. Comparaison des trois échelles :
# `controle/echelle-tourelles.png`. Deux nombres à retoucher si ça bouge.
#
# ⚠ LES ARTILLERIES SONT À 2,4, PAS À 1,6, ET CE N'EST PAS UNE INCOHÉRENCE.
# Leur socle est une coque à chenilles, plus haute que large et conditionnée à
# 85 % : le logement n'y vaut que 31 à 35 % de la coque, contre 45 % sur le
# socle carré. À échelle égale leur tourelle sortait donc nettement plus petite
# que celle d'une défense de contact, alors que ce sont les pièces à longue
# portée. +50 % demandé le 05/09 après lecture de la planche à 40 px.
ECHELLE = {'contact': 1.6, 'artillerie': 2.4}
ARTILLERIES = ('faucheuse', 'mortier', 'harpon')

# socle → tourelle qui s'y pose.
COUPLES = {
    'socle_def_j_casemate': 'def_j_casemate',
    'socle_def_j_creneau': 'def_j_creneau',
    'socle_def_j_batterie': 'def_j_batterie',
    'socle_def_j_faucheuse': 'def_j_faucheuse',
    'socle_def_j_mortier': 'def_j_mortier',
    'socle_def_j_harpon': 'def_j_harpon',
}


def charger(nom):
    a = np.array(Image.open(os.path.join(SRC, nom + '.png')).convert('RGB')).astype(int)
    r, g, b = a[..., 0], a[..., 1], a[..., 2]
    m = ~((g < 100) & (r > 110) & (b > 110) & (np.abs(r - b) < 90))
    return a, m


def pivot(m):
    """Le centre et le diamètre de l'embase — le plus grand disque inscrit.

    ⚠⚠ L'HYPOTHÈSE DU CYLINDRE VU DE BIAIS EST TOMBÉE AVEC LES DESSINS — lot
    ANCRES-ZÉNITH, 19/09. Cette fonction disait : « une embase est un cylindre vu
    de biais, sa silhouette atteint sa largeur maximale sur une bande, le pivot
    est sur la PREMIÈRE ligne de cette bande ». Vrai des douze tourelles à 75°,
    faux des douze tourelles zénithales : vu de dessus il n'y a plus de bande,
    et la largeur maximale est atteinte là où le dessin est le plus large — qui
    n'est pas l'embase. Mesuré sur `def_o_harpon` : la bande était la rangée
    de missiles, D = 858 pour une embase de 390, soit ×2,2 ; le pivot rendu
    tombait à 157 px au-dessus du disque. Sur les cinq autres de l'Ouvrage, la
    première ligne de la bande d'un DISQUE est à 0,1 D au-dessus de son centre —
    d'où les +3 à +12 % relevés par le brief.

    EN ZÉNITHAL, L'EMBASE EST LE PLUS GRAND DISQUE INSCRIT DANS LA SILHOUETTE :
    son centre est le maximum de la transformée de distance, son diamètre deux
    fois cette distance. Les canons, rampes et ailerons sont des appendices
    plus minces que l'embase et ne le déplacent pas. Vérifié sur les douze
    dessins, disque tracé sur chacun (`rapports/`, planche du lot) : les six de
    l'Ouvrage l'épousent, les six du joueur trouvent la plaque sous les canons,
    et les deux harpons ignorent leurs rampes.

    ⚠ LE MAXIMUM EST UN PLATEAU, PAS UN POINT. Sur une plaque plus large que
    haute, ou sur une embase ovale, la distance maximale est atteinte sur un
    SEGMENT : `def_o_mortier` la tient de y = 614 à 695 pour un anneau centré
    vers 655, et `argmax` rendrait le premier pixel, 40 px trop haut. On prend
    donc le centre des pixels à moins d'un centième du rayon du maximum — au
    moins un pixel : à un demi-pixel le plateau ne fait que 14 px et penche
    encore (640) ; à un centième il en fait 86 et tombe à 654. Coordonnées en
    bords de pixel, comme avant.

    ⚠ UNE APPROXIMATION ASSUMÉE : sur une embase polygonale ou à brides, le
    disque inscrit est l'apothème, un peu sous la largeur du plateau — 606 pour
    627 sur `def_o_casemate`, 3 %. Pour l'Ouvrage c'est sans effet sur le
    rendu, `echelle` étant calée sur le carré du joueur ; pour le joueur c'est
    la définition retenue : le plus grand disque qui tienne sous le corps.

    ⚠⚠ LE 75° N'A PLUS DE LECTEUR ICI, ET IL N'EST PAS PERDU. Les douze
    tourelles de défense sont zénithales depuis ce lot ; la règle du tambour vit
    encore, inchangée, dans `tools/ancres-blindes.py:pivot`, pour les dix
    tourelles de blindé qui restent à 75°. Aucun discriminant automatique
    75°/zénithal n'a tenu la mesure — cinq ont été essayés sur les vingt-quatre
    planches, voir le rapport —, donc deux fonctions, une par géométrie, chacune
    avec ses lecteurs. Le montage T1 du lot rejoue cet outil sur les sources 75°
    écartées AVEC la règle du tambour et retrouve le JSON d'avant à l'octet.
    """
    dist = nd.distance_transform_edt(m)
    rayon = float(dist.max())
    ys, xs = np.where(dist >= rayon - max(1.0, 0.01 * rayon))
    return float(xs.mean()) + 0.5, float(ys.mean()) + 0.5, int(round(2 * rayon))


def cote_du_carre(m, px, py):
    """Le côté du carré de rotation : deux fois la distance pivot → pixel le plus loin.

    ⚠⚠ IL SE MESURE, IL NE SE PRODUIT PLUS. La version livrée de cet outil
    écrivait aussi le PNG recentré, dans `art/sources/centrees/`. Deux raisons de
    ne pas le garder : `art/sources/` ne porte que des ORIGINAUX — « rien n'y est
    un produit, tout y est un original » (CLAUDE.md §2) — et surtout la mesure
    dit que le geste n'a rien à faire. Les six planches livrées sont DÉJÀ carrées
    et déjà centrées sur leur pivot : `def_j_casemate` mesure 720 × 720 pour un
    carré recalculé de 720, et les onze tourelles des deux familles sortent
    « OUI » au même contrôle. Le recentrage est donc idempotent, et le seul
    nombre qui sert en aval est ce côté-ci.

    ⚠ CE QUI DISPARAÎT AVEC LE PNG : le contrôle « aucun pixel perdu au
    recentrage » que l'outil livré faisait par assertion. Il n'a plus d'objet —
    on ne recentre plus rien. Ce qui le remplace est plus fort et vit dans
    `test/sprite.test.js` : le sprite est carré, et aucun pixel opaque n'est plus
    loin du centre que la moitié du côté.

    ⚠⚠ CE QUI PRÉCÈDE ÉTAIT VRAI DES PLANCHES À 75° ET EST FAUX DES PLANCHES
    ZÉNITHALES — lot ANCRES-ZÉNITH, 19/09, mesuré sur les douze. Elles sont
    carrées (1 254 ou 1 024) mais PAS centrées sur leur pivot : l'embase est de
    21 px (`def_o_casemate`) à 225 px (`def_j_mortier`) SOUS le centre du
    fichier, et le carré de rotation dépasse le côté du fichier sur dix d'entre
    elles (1 430 à 1 638 pour 1 254 chez le joueur). Le mode `carre` de
    `joueur_v2.py` — « aucun recadrage, la planche EST le sprite » — ne peut donc
    pas leur être appliqué tel quel : le lot de conditionnement devra RECENTRER
    chaque tourelle sur le pivot que cette fonction-ci reçoit et agrandir la
    toile au côté qu'elle rend, sans quoi la tourelle décrira un cercle de 20 à
    225 px autour de son embase en tournant. Ce lot-ci MESURE ; il ne
    conditionne pas (brief §5).
    """
    ys, xs = np.where(m)
    rayon = int(np.ceil(np.sqrt(((xs - px) ** 2 + (ys - py) ** 2).max())))
    return 2 * rayon


# ---------------------------------------------------------------------------
# De l'ancre MESURÉE à l'ancre que le rendu emploie
# ---------------------------------------------------------------------------
#
# ⚠⚠ DEUX RÉFÉRENTIELS, ET LES CONFONDRE FAIT DEUX FOIS LA TAILLE DE LA CASE.
# `chassis.py:ancre` mesure sur le DESSIN : `diametre_pct` est un pourcentage de
# la LARGEUR de la pièce, `x_pct` et `y_pct` des pourcentages de sa largeur et de
# sa hauteur. Le rendu, lui, ne connaît que le côté de la case. Or la pièce
# n'occupe pas la case entière : `recadrer` porte sa PLUS GRANDE dimension à
# `emprise / 32`, et l'autre suit le rapport du dessin.
#
# Prendre les nombres mesurés pour des pourcentages de case — ce que faisait le
# rendu du lot 8, et ce que la formule du brief reconduit — multiplie le carré
# par `32 / largeur_de_la_pièce`. Mesuré sur la v2 : le carré du Chasseur ferait
# **131,9 pixels de case sur 64**, soit deux fois la case ; celui de la Faucheuse
# 113,8. La conversion n'est donc pas un raffinement, c'est ce qui rend le
# montage possible.
#
# ⚠ ET C'EST LA FORMULE QUI REND LES PLAFONDS DU BRIEF. Emprise maximale à
# laquelle l'union coque + carré tient encore dans la case : **23,6 pour le
# Chasseur, 31,5 pour le Percheron, 32,0 pour les trois autres** — les trois
# nombres que le brief donne comme mesurés. La formule littérale du brief, elle,
# ne dépend pas de l'emprise et ne peut donc rendre aucun plafond.
#
# Les trois nombres publiés sont donc, en pourcentage de la CASE :
#   cote_case_pct  le côté du carré à dessiner
#   dx_case_pct    le décalage du centre du carré, horizontal
#   dy_case_pct    le décalage du centre du carré, vertical


def boite_dans_la_case(m, emprise):
    """La boîte de la pièce conditionnée, en pourcentage de la case.

    `recadrer` porte la plus grande dimension du sujet à `emprise / 32` de la
    case et centre la boîte ; l'autre dimension suit le rapport du dessin.
    """
    ys, xs = np.where(m)
    W = xs.max() - xs.min() + 1
    H = ys.max() - ys.min() + 1
    grand = max(W, H)
    return 100.0 * emprise / 32 * W / grand, 100.0 * emprise / 32 * H / grand


def ancre_de_case(anc, m, emprise, tour):
    """Les trois nombres du rendu, dérivés de l'ancre mesurée."""
    lc, hc = boite_dans_la_case(m, emprise)
    return {
        'cote_case_pct': round(lc * anc['diametre_pct'] / 100
                               * tour['cote_pct_embase'] / 100 * tour['echelle'], 2),
        'dx_case_pct': round(lc * anc['x_pct'] / 100, 2),
        'dy_case_pct': round(hc * anc['y_pct'] / 100, 2),
    }


def main():
    socles, tourelles, rapport = {}, {}, []
    emprises = emprise_par_sprite()
    for socle, tourelle in COUPLES.items():
        a, m = charger(socle)
        trouve = ancre(a, m)
        if trouve is None:
            raise AssertionError(f'{socle} : aucun logement détecté')
        socles[socle] = trouve

        _, mt = charger(tourelle)
        px, py, D = pivot(mt)
        cote = cote_du_carre(mt, px, py)

        # ⚠ LE RATIO N'EST PAS COSMÉTIQUE. Le sprite d'une tourelle n'est plus
        # collé sur son embase : il porte la marge de rotation, et cette marge
        # va de ×1,43 à ×2,09 selon la tourelle. Sans ce nombre, une tourelle
        # dessinée « au diamètre du logement » aurait son embase deux fois trop
        # petite, et pas du même facteur d'une tourelle à l'autre.
        famille = 'artillerie' if tourelle.split('_')[-1] in ARTILLERIES else 'contact'
        tourelles[tourelle] = {'cote_pct_embase': round(100 * cote / D, 1),
                               'echelle': ECHELLE[famille]}
        socles[socle].update(ancre_de_case(trouve, m, emprises[socle],
                                           tourelles[tourelle]))

        ys, xs = np.where(m)
        W = xs.max() - xs.min() + 1
        d_px = trouve['diametre_pct'] / 100 * W
        rapport.append((tourelle, D, cote, round(cote / D, 3), trouve, round(d_px)))

    with open(ANCRES, 'w', encoding='utf-8') as f:
        json.dump({'socles': socles, 'tourelles': tourelles},
                  f, ensure_ascii=False, indent=2, sort_keys=True)
        f.write('\n')

    print(f'{"tourelle":<18}{"embase":>8}{"carré":>8}{"ratio":>8}   logement du socle')
    for nom, D, cote, ratio, anc, d_px in rapport:
        print(f'{nom:<18}{D:>8}{cote:>8}{ratio:>8}   '
              f"{anc['diametre_pct']}% = {d_px} px, x {anc['x_pct']}%, y {anc['y_pct']}%")
    print()
    print(f'{"socle":<24}{"côté/case":>11}{"dx/case":>9}{"dy/case":>9}')
    for socle in COUPLES:
        v = socles[socle]
        print(f"{socle:<24}{v['cote_case_pct']:>10.2f}%{v['dx_case_pct']:>8.2f}%"
              f"{v['dy_case_pct']:>8.2f}%")
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: src/data/ancres-defense.js
// Où se pose la tourelle sur le socle d'une défense — LES DEUX CAMPS.
//
// ⚠⚠ TRANSCRIPTION À LA MAIN DE `art/sprites/ancres-defense.json` ET DE
// `ancres-defense-ouvrage.json`, ET UN TEST LES CONFRONTE. Les deux JSON sont
// produits par `tools/ancres-defense.py` et `tools/ancres-ouvrage.py`, qui
// MESURENT le logement de tourelle sur l'image ; ils ne peuvent pas entrer dans
// le livrable — `tools/build.js` n'inline que des images, et `render/scene.js`
// ne lit aucun fichier. La transcription est donc la seule voie, et une
// transcription qui ne se confronte pas à sa source est une copie qui vieillit.
// Le test compare les clés ET les valeurs SIGNÉES, dans les deux sens, contre
// l'UNION des deux fichiers.
//
// ⚠⚠ UNE SEULE TABLE POUR LES DEUX CAMPS, ET C'EST CE QUI ÉVITE UN `=== 'o'`.
// Elle est indexée par NOM DE SPRITE, et `socle_def_o_casemate` ne peut pas
// entrer en collision avec `socle_def_j_casemate` : la fusion est sûre par
// construction. Voir `src/data/ancres-blindes.js` pour le motif complet — c'est
// le même, et il est écrit une fois là-bas.
//
// ⚠⚠ ET C'EST LA SEULE CHOSE QUI FAIT TOURNER LES SIX DÉFENSES DE L'OUVRAGE.
// `couchesDeLaDefense` n'a pas changé d'une ligne : elle écrivait déjà
// `socle_def_${c}_${d.id}` et lisait `ANCRES_DEFENSE[socle] ?? null`. Tant que
// la table ne portait que le joueur, le `?? null` rendait `null` pour l'Ouvrage
// et `dessinerCouches` posait la pièce sur la case entière. L'arrivée de ces six
// clés suffit à la faire tourner. Le discriminant est la DONNÉE, jamais le camp.
//
// ⚠⚠ DEUX RÉFÉRENTIELS COHABITENT ICI, ET LES CONFONDRE FAIT DEUX FOIS LA
// TAILLE DE LA CASE. Les trois premiers nombres sont MESURÉS sur le dessin, en
// pourcentage de la PIÈCE ; les trois derniers sont ceux que le rendu emploie,
// en pourcentage de la CASE. La pièce n'occupe pas la case entière —
// `recadrer` porte sa plus grande dimension à `emprise / 32` — donc prendre les
// premiers pour les seconds multiplie le carré de la tourelle par
// `32 / largeur_de_la_pièce`. Mesuré sur la v2 : le carré du Chasseur ferait
// **131,9 pixels de case sur 64**, deux fois la case, et celui de la Faucheuse
// 113,8.
//
//   diametre_pct   diamètre du logement, en % de la largeur de la pièce
//   x_pct, y_pct   décalage du centre du logement, en % de sa largeur / hauteur
//   cote_case_pct  le côté du carré de tourelle À DESSINER, en % de la case
//   dx_case_pct    décalage du centre de ce carré, en % de la case, vers la droite
//   dy_case_pct    décalage du centre de ce carré, en % de la case, vers le BAS
//   mesure         le logement a-t-il été trouvé sur l'image, ou estimé
//
// ⚠ `cote_case_pct` PORTE DÉJÀ L'ÉCHELLE ET LA MARGE DE ROTATION. Il vaut
// `largeur_de_la_pièce_dans_la_case × diametre_pct/100 × cote_pct_embase/100 ×
// echelle` : le rendu n'a donc rien à multiplier, et les deux nombres de la
// table des tourelles ci-dessous sont là pour dire D'OÙ il vient, pas pour être
// relus au dessin.
//
// ⚠⚠ LA TABLE EST MIXTE DEPUIS LE LOT ANCRES-ZÉNITH (19/09), ET C'EST VOULU :
// NEUF SOCLES MESURÉS SUR DES DESSINS ZÉNITHAUX, TROIS SUR DES DESSINS À 75°.
// Les six socles de l'Ouvrage et les trois socles de tourelle du joueur
// (`casemate`, `creneau`, `batterie`) ont été redessinés vus de dessus, et leur
// logement est AU CENTRE de la pièce — `y_pct` de −0,2 à −1,2 là où il valait
// −17,6 à −35,3. Les trois socles d'artillerie du joueur (`faucheuse`,
// `mortier`, `harpon`) n'ont PAS été redessinés — arbitrage d'Ethan du 19/09,
// « si j'ai pas modifié, c'est qu'il n'y a pas besoin » —, et l'ancre décrit le
// socle : le socle n'a pas bougé, l'ancre non plus (−11,1 · −10,9 · −11,8).
// **Ne pas « harmoniser » ces trois-là** : les relancer dans le détecteur
// réparé rendrait d'ailleurs exactement les mêmes nombres (T1 du lot, au bit
// sur les 30 pièces 75° du dépôt) ; seul un socle redessiné les changera, et
// `test/ancres-zenith.test.js` le dira. ⚠ Leur `cote_case_pct`, lui, a bougé
// avec leur TOURELLE, qui est redessinée : 111,95 → 146,89, 104,96 → 159,17,
// 96,53 → 180,01.
//
// ⚠⚠ ET DEUX DÉTECTEURS ONT ÉTÉ RÉPARÉS POUR CES DESSINS-LÀ, SANS DÉPLACER LE
// 75°. `chassis.ancre` rendait 82,7 % et +7,3 sur le créneau de l'Ouvrage — une
// tache plus haute que large, le logement plus le module inférieur du socle —
// et rejetait les trois marcheurs ; `ancres-defense.pivot` prenait la rangée de
// missiles du Harpon pour son embase (D 858 pour 390). Le détail, mesure par
// mesure, est dans `tools/chassis.py`, `tools/ancres-defense.py` et
// `rapports/RAPPORT-lotANCRES-ZENITH.md`.
//
// ⚠⚠ LE CARRÉ DÉBORDE DE LA CASE SUR DIX SOCLES SUR DOUZE, ET LE MOTIF A CHANGÉ.
// Au 08/09 il débordait sur les douze — 58 à 65 % de demi-case chez le joueur,
// 70 à 84 à l'Ouvrage — parce que le logement était haut sur la pièce. Le
// logement est centré maintenant ; ce qui déborde, c'est le carré lui-même :
// les tourelles zénithales portent des canons et des rampes deux à trois fois
// plus longs que leur embase (`cote_pct_embase` 246 à 336 chez le joueur,
// contre 143 à 209 avant), donc à `echelle` inchangée le carré fait **96 à
// 180 % de la case** — le Créneau, seul, tient dans la sienne (48,98 et 48,18),
// le Harpon joueur atteint 99,97 % de demi-case. **`echelle` a été choisie à
// l'œil sur les dessins à 75° ; la rejuger sur les zénithaux est un arbitrage
// d'Ethan, au lot de conditionnement.** Un test borne les deux camps
// séparément et nomme le pire de chaque côté au centième.
//
// ⚠ ET LA v1 FAISAIT PIRE, ce qui met ces chiffres en perspective : elle
// dessinait la tourelle sur la case ENTIÈRE, `sprite(famille, nom, x, y, t, t)`,
// son canon atteignant le bord par construction.

/**
 * @typedef {{ cote_case_pct: number, diametre_pct: number, dx_case_pct: number,
 *            dy_case_pct: number, mesure: boolean, x_pct: number, y_pct: number }} Ancre
 * @type {Record<string, Ancre>}
 */
export const ANCRES_DEFENSE = {
  socle_def_j_batterie: { cote_case_pct: 130.2, diametre_pct: 26.9, dx_case_pct: -0.0, dy_case_pct: -1.06, mesure: true, x_pct: -0.0, y_pct: -1.2 },
  socle_def_j_casemate: { cote_case_pct: 107.19, diametre_pct: 26.9, dx_case_pct: -0.0, dy_case_pct: -1.06, mesure: true, x_pct: -0.0, y_pct: -1.2 },
  socle_def_j_creneau: { cote_case_pct: 95.84, diametre_pct: 26.9, dx_case_pct: -0.0, dy_case_pct: -1.06, mesure: true, x_pct: -0.0, y_pct: -1.2 },
  socle_def_j_faucheuse: { cote_case_pct: 146.89, diametre_pct: 35.4, dx_case_pct: -0.0, dy_case_pct: -9.37, mesure: true, x_pct: -0.0, y_pct: -11.1 },
  socle_def_j_harpon: { cote_case_pct: 180.01, diametre_pct: 35.3, dx_case_pct: -0.07, dy_case_pct: -9.96, mesure: true, x_pct: -0.1, y_pct: -11.8 },
  socle_def_j_mortier: { cote_case_pct: 159.17, diametre_pct: 31.2, dx_case_pct: -0.0, dy_case_pct: -9.2, mesure: true, x_pct: -0.0, y_pct: -10.9 },
  socle_def_o_batterie: { cote_case_pct: 130.18, diametre_pct: 55.1, dx_case_pct: 0.0, dy_case_pct: -0.18, mesure: true, x_pct: 0.0, y_pct: -0.2 },
  socle_def_o_casemate: { cote_case_pct: 107.19, diametre_pct: 55.0, dx_case_pct: -0.0, dy_case_pct: -0.36, mesure: true, x_pct: -0.0, y_pct: -0.4 },
  socle_def_o_creneau: { cote_case_pct: 95.82, diametre_pct: 55.0, dx_case_pct: -0.0, dy_case_pct: -0.27, mesure: true, x_pct: -0.0, y_pct: -0.3 },
  socle_def_o_faucheuse: { cote_case_pct: 146.88, diametre_pct: 24.2, dx_case_pct: -0.0, dy_case_pct: -0.47, mesure: true, x_pct: -0.0, y_pct: -0.6 },
  socle_def_o_harpon: { cote_case_pct: 179.99, diametre_pct: 22.1, dx_case_pct: -0.0, dy_case_pct: -0.56, mesure: true, x_pct: -0.0, y_pct: -0.7 },
  socle_def_o_mortier: { cote_case_pct: 159.18, diametre_pct: 20.4, dx_case_pct: -0.0, dy_case_pct: -0.62, mesure: true, x_pct: -0.0, y_pct: -0.8 },
};

/**
 * D'où vient `cote_case_pct` — la marge de rotation du sprite, et le facteur
 * de lisibilité. Aucun des deux n'est relu au dessin ; ils sont ici pour que le
 * nombre du dessus ne soit pas un nombre tombé du ciel.
 *
 * ⚠ `cote_pct_embase` est le rapport du carré du sprite à son embase — il va de
 * ×1,48 à ×3,34 selon la tourelle depuis le lot ANCRES-ZÉNITH (×1,43 à ×2,21 sur
 * les dessins à 75°), parce qu'un tube long demande plus de marge pour tourner
 * sans se rogner, et que les dessins zénithaux ont des tubes plus longs. Chez le
 * JOUEUR, `echelle` vaut 1,6 pour les trois tourelles de contact et 2,4 pour les
 * trois artilleries, CHOISI À L'ŒIL à 40 px sur les dessins à 75° : à ×1,0 le
 * canon disparaît et il ne reste qu'un anneau de couleur.
 *
 * ⚠ EN ZÉNITHAL, L'EMBASE EST LE PLUS GRAND DISQUE INSCRIT DANS LA SILHOUETTE
 * — `tools/ancres-defense.py:pivot`. Sur une plaque plus large que haute
 * (`def_j_batterie`, 742 × 370 px) c'est la hauteur qui compte, d'où un ratio
 * de 3,34 ; prendre la largeur de la plaque aurait donné 1,85 et un carré de
 * 72 % au lieu de 130. C'est la définition retenue — le plus grand disque qui
 * tienne sous le corps dessiné — et elle est écrite là où elle se mesure.
 *
 * ⚠⚠ LES SIX ÉCHELLES DE L'OUVRAGE SONT CALCULÉES, PAS CHOISIES — DÉCISION
 * D'ETHAN DU 07/09. Chacune aligne le carré de la tourelle de l'Ouvrage sur
 * celui de son homologue du joueur, en pourcentage de case, si bien que les deux
 * camps se lisent à la même taille. Elle se RECALCULE quand un dessin change, au
 * lieu d'être une constante à re-arbitrer — voir `src/data/ancres-blindes.js`,
 * qui porte la formule. Vérifiable dans la table ci-dessus : Faucheuse 146,89
 * contre 146,88, Batterie 130,20 contre 130,18, Casemate 107,19 des deux côtés
 * (au 08/09 : 111,95 · 93,82 / 93,83 · 93,36 / 93,35). ⚠ Le Créneau de
 * l'Ouvrage passe SOUS 1 (0,992) : son embase est plus petite que son carré ne
 * le demande, et c'est le calage qui le dit, pas un choix.
 */
export const TOURELLES_DEFENSE = {
  def_j_batterie: { cote_pct_embase: 333.8, echelle: 1.6 },
  def_j_casemate: { cote_pct_embase: 274.8, echelle: 1.6 },
  def_j_creneau: { cote_pct_embase: 245.7, echelle: 1.6 },
  def_j_faucheuse: { cote_pct_embase: 274.5, echelle: 2.4 },
  def_j_harpon: { cote_pct_embase: 304.9, echelle: 2.4 },
  def_j_mortier: { cote_pct_embase: 303.3, echelle: 2.4 },
  def_o_batterie: { cote_pct_embase: 205.6, echelle: 1.268 },
  def_o_casemate: { cote_pct_embase: 147.5, echelle: 1.458 },
  def_o_creneau: { cote_pct_embase: 193.6, echelle: 0.993 },
  def_o_faucheuse: { cote_pct_embase: 325.5, echelle: 2.21 },
  def_o_harpon: { cote_pct_embase: 270.3, echelle: 3.571 },
  def_o_mortier: { cote_pct_embase: 282.9, echelle: 3.269 },
};
